### PR TITLE
Additional catalog client functionality for cache eviction

### PR DESCRIPTION
**Pull request by MorpheusXAUT**
# TL;DR
Added extra methods for accessing and removing artifact data to catalog client.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [ ] Unit tests added
 - [X] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
The new `ByArtifactTag` methods were added to allow for easier access to reservations from flyteadmin (which retrieves the artifact tag from its stored metadata without having to access the stored artifact to generate a `catalog.Key`).
Additional functionality for deleting artifacts has been added as well.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2867

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by MorpheusXAUT** [comment]
Rebased onto current `master` branch.
